### ISSUE TITLE
SGE Changes Tracking

### ISSUE THREAD

**Issue by tatarsky** [frazer-lab/cluster#24]
I will be documenting when I make changes to the SGE config as we move toward a first goal of queue and settings config here. I will also attempt to get into the Wiki the details of such changes usages.

**Comment by tatarsky** [comment]
Changed default qsub priority to 100 from 0. This allows a user to `qalter` their own priorities on queued obs later or on the qsub command line without an SGE manager intervention. Details coming to wiki when I get a moment. Goal is to allow user to re-order their own priorities within reason once queued.

**Comment by tatarsky** [comment]
Initial default qsub memory target will be 4GB. Memory as a consumable is NOT enabled yet. But noting this selection from Skype call.

**Comment by tatarsky** [comment]
Due to the priority request the two high memory nodes will be removed from the all.q (default) in a few minutes. Please confirm the name of the desired long running queue. I believe it was `long`

**Comment by nariai** [comment]
long queue is fine. thanks!

**Comment by tatarsky** [comment]
Nodes 15 and 16 will shortly not be in the default queue (all.q). 

Do you want just those two nodes in the long running queue or do you want all the nodes to be able to run some number of long.q jobs as well if they are idle?

**Comment by tatarsky** [comment]
Memory as a consumable will be prepared today but NOT activated as @hurleyLi is running I believe a stack of jobs. I will be oversubscribing ram on the nodes by 20% to start. The default qsub if non-specified will be 4GB.

**Comment by nariai** [comment]
Let's do the first option (nodes 15 and 16 as the long queue).
We want to use the long queue as the high memory requiring jobs.

**Comment by tatarsky** [comment]
When the jobs on node 15 and node 16 are complete I will test that the long.q is correct and ready for use and add the usage statements to the wiki.

What was the other queue you wanted?

**Comment by nariai** [comment]
Also, as discussed, please make 128 cores (four nodes amount) as the short queue (2 days maximum).
In this case, let all the nodes be able to run the short queue as well if they are idle.

The remaining nodes (10 nodes) will be the week queue (7 days maximum).
In this case, if the high memory nodes (2 nodes) are idle, then these nodes can be
used for the week queue.

**Comment by tatarsky** [comment]
Noted.

**Comment by tatarsky** [comment]
What queue do you want as default if they don't specify it. I assume `week`

**Comment by nariai** [comment]
Yes, let's make the week queue as a default.

**Comment by tatarsky** [comment]
Week queue for simplicity is going to simply be the all.q for now. I will set the maximum run time to one week after @hurleyLi jobs are done.

**Comment by hurleyLi** [comment]
I'll stop submitting jobs for now then.

**Comment by tatarsky** [comment]
No, thats not been part of any thing I've recommended. I was told you should be pushing jobs through! My efforts here are to NOT conflict with that as a stated goal from Kelly.

**Comment by hurleyLi** [comment]
Sorry I misunderstood.. But if I'm gonna continuously running those jobs, they'll probably be finished in at least a week. So you'll activate the changes at that time? 
Another thing I noticed is there're two jobs running on n16 now, although both of them are specified as using 32 cores.

**Comment by tatarsky** [comment]
I've been told your jobs are the priority. If my changes are going to interfere with that I will leave the changes until when you are done. 

However, yes, you are correct on that n16 matter. Which is a config item I need to finish when I believed getting that queue was my priority. 

I might as well finish it if you can hold a moment.

**Comment by tatarsky** [comment]
I've disabled that queue for now. Shouldn't happen again and will let your jobs finish.

**Comment by tatarsky** [comment]
Basically and I could use some clarification still. Are people waiting to run jobs besides @hurleyLi ?
If so, why? Just because we don't have named queues yet? I am trying to honor a request to not cause issues with his jobs but configuring a scheduler while people are using it is not really my favorite thing to do. However, there is nothing preventing anyone from issuing jobs to the default queue. 

If @hurleyLi reduces his slot count a bit as noted in #34 there will even be a place to run such jobs.But I defer to his goals in getting things done for those numbers.

**Comment by hurleyLi** [comment]
Hi Paul, how long does it take for you to activate/implement those changes?

**Comment by tatarsky** [comment]
I have the first pass ready for some testing to make sure what has been described is what will occur. I am not however changing the name of the all.q to week.q at the moment due to your jobs in said queue.

**Comment by tatarsky** [comment]
For example at the moment due to node 3 being idle I am testing the short queue.

**Comment by hurleyLi** [comment]
What I can do is I can kill the jobs and let you make the changes, coz I also want these to be implemented and tested sooner than later. I have to stop at some point anyway for you the activate these changes, if I don't stop, the whole process after these jobs will last a month. 

Now I'm really just waiting 2-3 jobs to finish. 276, 279, 285. They'll probably finish around 2-3pm this afternoon. Do you think you can activate the changes this afternoon, so I can restart those jobs later today. And at the meantime, I can test some of my changes as mentioned in #34 

Does it sound a plan?

**Comment by tatarsky** [comment]
I will implement the changes when I see your jobs exit. Be aware your 2-3PM is my end of day. Afterhours monitoring is best effort which is why I normally do not make heavy changes on a Friday. 

But in the interests of moving this along and given I am mostly around this weekend I will attempt to implement the items above for the queues.

IMPORTANT question however: are you really ready for memory reservations? I can do that separately. But you would need to add the proper `-l h_vmem=XXG` to your jobs....

As we are outpacing my ability to document and configure we'd have to sort out any issues as we go.

**Comment by hurleyLi** [comment]
Do you prefer to do this on Monday? We can do this next week if it's best for you to monitor the cluster after these changes. I don't think other people are gonna use the cluster beside me

**Comment by tatarsky** [comment]
If that is truly the case I would prefer Monday. That way you will get uninterrupted use this weekend. While SGE is fairly direct, it periodically gets confused and I'd rather not scramble through config if I can do it during regular work hours.

**Comment by nariai** [comment]
That's fine for me. Let's change the config on Monday.

**Comment by tatarsky** [comment]
However as it does not interfere with jobs the short.q is ready for test if people would like. I have I believe the spirit of the requirements above defined but may need to adjust the assignment of up to 128 cores in a different way. (Right now it maxes out at 8 per node and we may want to float that a bit).

If you wish to test it you can use the following syntax.

```
qsub -l short (script)
```

There is a reason I will elaborate on for `-l short` as the method for this type of queue. 

Jobs in the short queue are killed with a SIGUSR1 after 48 hours as a warning. 

If you trap that signal you get another 10 minutes to shutdown gracefully as your reward for understanding how to trap signals.

**Comment by nariai** [comment]
It looks like the node fl-n-1-8 only accepts 8 jobs (8 slots in total).
Given that each node has 128 GB of mem in total, we want to run more jobs than 8 per node.
(Same for the node fl-n-1-6.)

Paul, when you change the config on Monday, is it possible to make the maximum number of jobs per node as 32 (given that the default mem_req is 4GB)?

Does this make sense to everybody? Any other ideas on the number of jobs per node?

Naoki

**Comment by cdeboever3** [comment]
Maximum of 32 jobs per node sounds good.

**Comment by tatarsky** [comment]
The short.q is currently limited to 8 jobs per node.

**Comment by tatarsky** [comment]
The method that is done will be changed to allow a total short.q run of 128 slots total on all nodes regardless of the per node count. But shortly. 32 slots per node is already the default for the all.q (week.q shortly) and also the long.q.

**Comment by hurleyLi** [comment]
Hi Paul, my jobs on node n8 are suddenly end with error. And it seems I can't write on scratch any more

```
fl-n-1-8: mkdir: cannot create directory `test': Read-only file system
```

I did a some test run earlier, and writing on scratch is a little bit faster than writing on luster, that's why I write all the temp files on scratch. But it's only running one job each node, I guess I'll just directly write to luster is I'm doing multiple jobs (4 maybe) on each node, right?

**Comment by tatarsky** [comment]
I would appreciate if issues would be opened for new items as a proper new issue. This may be a node problem.

**Comment by tatarsky** [comment]
The short.q was made capable of running 32 slots per node. I need to verify however that per your statement about that I have limited the short.q to a total running slot count of 128. That may not be correct at this moment in time.

**Comment by tatarsky** [comment]
@hurleyLi please refer to #39 for your node 8 issue. We appear to have a DOA drive.

**Comment by nariai** [comment]
Now I see that short.q@fl-n-1-6 is running with 21 slots.

**Comment by tatarsky** [comment]
Just to clarify to be sure, I show node 1-6 is fully loaded now with 32 slots. So I assume this was before you queued some other items. 

```
qstat|grep fl-n-1-6
    473 0.45617 insert_var nnariai      r     11/06/2015 16:41:49 short.q@fl-n-1-6                   4        
    474 0.45617 insert_var nnariai      r     11/06/2015 16:41:49 short.q@fl-n-1-6                   4        
    475 0.45617 insert_var nnariai      r     11/06/2015 16:41:49 short.q@fl-n-1-6                   4        
    476 0.45617 insert_var nnariai      r     11/06/2015 16:41:49 short.q@fl-n-1-6                   4        
    477 0.45617 insert_var nnariai      r     11/06/2015 16:41:49 short.q@fl-n-1-6                   4        
    478 0.45617 insert_var nnariai      r     11/06/2015 16:41:49 short.q@fl-n-1-6                   4        
    479 0.45617 insert_var nnariai      r     11/06/2015 16:41:49 short.q@fl-n-1-6                   4        
    480 0.45617 insert_var nnariai      r     11/06/2015 16:41:49 short.q@fl-n-1-6                   4        
```

Thats all 32 slots.

**Comment by nariai** [comment]
I was trying to optimize mysql queries from multiple compute nodes concurrently. Now it seems working great with just one slot per job.

**Comment by tatarsky** [comment]
Yep. That would be similar to what I see many people do. Usually you only reserve multiple cpus in SGE if you know your code is threaded or you are doing some MPI work.

**Comment by tatarsky** [comment]
I believe at some point people would like the "all" queue renamed "week" although we do not have to do that if you would just like the _rules_ for the week queue applied to the default all.q. (runtime one week and slots rules). There will be no "selector" for this queue as its the default. (Compared to short and long).

Also I can remove nodes 15 and 16 from that queue if you would like today but remain of the impression @hurleyLi is operating under a deadline and that his smaller core granularity of jobs is allowing others to proceed as well. 

Advise when you can.

**Comment by nariai** [comment]
As discussed, we hope to change the SGE setting (queue stratification and h_vmem setting) on Monday. This change will give us the opportunities to run other types of jobs in the short queue, while the BWA+GATK are running in the week queue and the long queue. We plan to do other types of analysis on the short queue after the changes are applied. Thanks!

**Comment by tatarsky** [comment]
OK. I will start the process. The main item once I turn on h_vmem is that if a job is not submitted with a 

```
-l h_vmem=XXG
```

The default is 4GB. And SGE will _kill_ the job if it goes over the default or what was requested.

**Comment by tatarsky** [comment]
Taking nodes 15 and 16 out of the default queue and enabling the long queue. This will require the following to run jobs on them:

```
qsub -l long (script)
```

**Comment by nariai** [comment]
> -l hvem=XXG

Noted. We hope that this change will make the cluster be more stable.

**Comment by tatarsky** [comment]
Generically I find memory reservations well worth it. I have activated the memory as a consumable on all queues. Existing jobs are not affected. You can see the default limit if you do a qlogin without a memory request as they are implemented as shell limits. 

```
$ ulimit -a
virtual memory          (kbytes, -v) 4194304
```

**Comment by tatarsky** [comment]
all.q will be disabled in a moment to allow it to drain. Then it will be deleted. This is just to implement a naming change to week.q as the default. The runtime limit on week.q was changed to 1 week with a 10 minute soft/hard grace period if you trap the signal.

**Comment by tatarsky** [comment]
You should notice now any job submitted without a selector (short|long) will go into the `week.q`. More details to wiki when I finish validation.

**Comment by nariai** [comment]
Can we start submitting jobs on the week queue now? Or should we hold on?

**Comment by tatarsky** [comment]
No go ahead.

**Comment by tatarsky** [comment]
Note clearly: no qsub arguments are required to submit to the week.q its default.

**Comment by nariai** [comment]
OK, thanks.

**Comment by tatarsky** [comment]
So the basic summary so far this morning:
1. Be sure to specify the level of ram your job requires with `-l h_vmem XXG`
2. The `short.q` which actually existed last week is selectable with `-l short` and will KILL your jobs with SIGUSR1 at 48 hours and SIGTERM at 48:10:00.
3. The long.q is selectable with `-l long` and is ONLY nodes 15 and 16 which are also the high memory nodes. The Short.q and the Week.q however can use nodes 15 and 16 if idle. But submitting to the long.q will ONLY run on nodes 15 and 16. (Make sure I understood you correctly here)
4. The default queue is the `week.q`. No selection is required to submit to it. Your jobs will be sent a SIGUSR1 at 168 hours (one week) and a SIGTERM at 168:10:00.

**Comment by tatarsky** [comment]
Documentation to follow if the above is correct and as I have time this morning.

**Comment by hurleyLi** [comment]
Hi Paul, do you need me to kill my jobs now or shortly?

**Comment by tatarsky** [comment]
No. Your jobs can run to completion just fine.

**Comment by tatarsky** [comment]
I simply disabled the all.q which allows existing jobs to finish. All NEW jobs be sure you understand the rules above.

**Comment by hurleyLi** [comment]
All right. And we can start to submit new jobs right?

**Comment by tatarsky** [comment]
That is correct.

**Comment by hurleyLi** [comment]
Great! Thanks Paul!

**Comment by tatarsky** [comment]
Let me know if the queues do not implement the goals above and we'll adjust as needed.

**Comment by tatarsky** [comment]
Man pages for SGE added BTW. They were left out of deploy by accident.

**Comment by cdeboever3** [comment]
To make sure I understand, If I submit:

-l long : only run on the two long nodes
-l week (default) : run on any node (long nodes only if idle), week time
limit
-l short : run on any node (long nodes only if idle), 48 hour time limit

Priority is long > short > week?

**Comment by tatarsky** [comment]
There is no `-l week`. I find having a default queue with a resource requirement to be more confusing than its worth (people forget and jobs sit there with no place to go). 

So the `week` queue requires no declared resource. 

The other two comments are correct for short and long. 

Priority at the moment is not defined but I will adjust if your statement is agreed to. Aka its currently first come first served.

**Comment by tatarsky** [comment]
And this will go into the wiki after I get some cycles. Monday always filled with entropy to fight.

**Comment by nariai** [comment]
To be clear, we want to make the priority:
long (two high mem nodes) > week (ten nodes: 320 cores + two high mem nodes if available)

And, the short queue as the separate from the priority above:
short (four nodes: 128 cores in total)

Is this possible?

**Comment by cdeboever3** [comment]
Sounds good. What do others think of the priority I suggested? I think it
will help make the short queue useful for testing.

**Comment by tatarsky** [comment]
I don't understand you mean by "separate from the priority above". Priorities involve the way the queued jobs are ordered when waiting for the next slot.

**Comment by tatarsky** [comment]
You've also stated that both the week and short queues can use any node given its idle. Once a job is running the concept of "priority" is gone. I assume you know what I mean...

**Comment by tatarsky** [comment]
But to be clear, I can give jobs submitted to "short" a bump or a decrement in that queue list. But I'm trying to clarify what you mean by "separate" as I can't really order things "separately".

**Comment by cdeboever3** [comment]
I think Naoki wants to limit "short" jobs to four nodes in the way that we
limit "long" jobs to two nodes. The short jobs would have higher priority
for those four nodes they are allowed to run on. If short nodes are idle,
week can use them (in the same way week can use the long nodes when idle).

**Comment by tatarsky** [comment]
That is not consistent with the earlier statements. 

"Also, as discussed, please make 128 cores (four nodes amount) as the short queue (2 days maximum).
In this case, let all the nodes be able to run the short queue as well if they are idle."

The second sentence implies that any node can run a short.q job. (If I understand it)

**Comment by tatarsky** [comment]
I am currently imposing I believe a queue limit of 128 slots total for short.q but they can be on ANY node.

**Comment by tatarsky** [comment]
BTW, I _think_ I have the short.q limited to 128 slots ;) When we get some open slots we should verify.

**Comment by tatarsky** [comment]
The key question in the second sentence is the meaning of "to run the short queue". Did that mean that week.q jobs can run on those hosts or that all hosts are capable of running short.q jobs.

**Comment by tatarsky** [comment]
BTW: If Git issue based untangling of requirements is too complex, I'm happy to Skype. Just tell me when ;)

I recognize there are subtleties to these concepts that are sometimes best understood verbally. I can do many things, but I want to make sure I understand what you are asking for.

**Comment by nariai** [comment]
Yes, let's discuss over the skype on this.

My argument was that 128 cores are specifically reserved for the short queue, and let no long gueue or week queue can use them, even if they are idle. The reason we want to impose this is that once the week queue or long queue use up all the slots (when there is no short queue submitted at that time), we have to wait for more than one week to start the short queue, if want to do the quick analysis in the short queue.

Does this make sense to everybody?

**Comment by hurleyLi** [comment]
This is what I understand about the current configuration based on the above conversation:
1) jobs submitted as  "-l short" can run on any node (including n15 and n16), until they hit a total slot limit of 128, then they'll just wait on the queue, even if there're empty slots.
2) jobs submitted as "-l long" can only run on n15 and n16
3) jobs submitted as default queue can run on any node (including n15 and n16), until they hit a total slot limit of 32 X 12 = 384 slots, then the rest of the jobs will wait on the queue, even if there're still empty slots. 
4) The priority now is first come first serve, but I think Naoki wants higher priority for jobs submitted as "-l short", so that these jobs will fill up the 128 slots first.

**Comment by tatarsky** [comment]
Currently @hurleyLi is mostly correct.

However, it sounds like the desire is for a group of four nodes to be dedicated to short.q jobs only. 

I can do that. But please confirm. That will mean once implemented those four nodes will ONLY run jobs if `-l short` is specified. I would recommend you use whole nodes instead of some slots per node or you will get in the way of jobs wanting 32 cores.

**Comment by tatarsky** [comment]
Also happy to Skype.

**Comment by cdeboever3** [comment]
Perhaps we can discuss amongst ourselves a bit and come to Paul with a more
unified vision on Wednesday's call. Based on the lab's historical work
patterns, I don't like the this:

> My argument was that 128 cores are specifically reserved for the short
> queue, and let no long gueue or week queue can use them, even if they are
> idle.

**Comment by nariai** [comment]
Thanks @hurleyLi. If this setting is possible, then I'm happy with this. But we will talk about this in the group and get back to you.

**Comment by tatarsky** [comment]
I believe I have what @hurleyLi describes at the moment but am willing to refine. I understand your goals just there are a couple of ways to reach them. I find generically if you dedicate nodes to queues you end up with wasted cycles. Nesting the queues and limiting either total _user_ job count or slot count is usually more efficient.

**Comment by tatarsky** [comment]
And please note that until the remaining "all.q" jobs exit, there may be inconsistency....

**Comment by nariai** [comment]
To be clear, if what @hurleyLi describes can be accomplished, I don't think we need the following:

> four nodes to be dedicated to short.q jobs only.

But anyway we will talk in the group and get back to you.

**Comment by tatarsky** [comment]
I believe what @hurleyLi describes is both possible and what I currently have implemented. I will know a bit more precisely when the all.q remaining jobs are gone if what are called "queue slot quotas" are working as I believe them to be.

**Comment by tatarsky** [comment]
The proof will be jobs in the week.q being queued when the total slot count of the week.q > 384.
And similar with short.q > 128.

**Comment by tatarsky** [comment]
(regardless of node)

**Comment by nariai** [comment]
Chis, please also remember that given the setting as per @hurleyLi, we can always fill-up the short queue by any jobs (this can be BWA+GATK as well). I just don't want to wait the short jobs in the short queue for more than a week.

**Comment by tatarsky** [comment]
Note I believe quotas on queues as listed above is working. I have temporarily limited the week.q to 256 slots so I could see the results of the limit a bit earlier. (Given some non-limited all.q jobs still in the mix)

@hurleyLi week.q jobs are being held due to that limit and the short.q jobs of @nariai just got n15 slots. Which I believe to be the intent of some of the discussion above. 

I will move the week.q quota up slowly and note a very useful command for folks to see why a job is queued:

```
   1282 0.55617 WGS_alignA hel070       qw    11/09/2015 08:14:24                                   10        

qstat -explain a -j 1282

(lots of output)
```

But in that list is this line:

```
                            cannot run because it exceeds limit "/////" in rule "weeklimit/1"
```

Which is the quota limit on the slots maximum on any node for week.q.

**Comment by tatarsky** [comment]
And I have now raised the week.q quota to the desired 384 slots. Lets see how we do with this method in terms of keeping the run queue nice and high on all the nodes and see if short.q jobs get into the run list OK.

**Comment by cdeboever3** [comment]
How does the h_vmem option work with shared memory? For instance, if I run
STAR and load the genome index into shared memory, do I need to request
space for the genome? I'd like to run STAR by loading the genome into
shared memory if it's not already there and having subsequent jobs use the
genome stored in shared memory. I'll remove all the shared memory genomes
after alignments are done.

**Comment by tatarsky** [comment]
If I recall correctly shared memory is "outside" the h_vmem concept as h_vmem is just a UNIX shell limit in implementation and I don't believe that includes shared memory. We may need to alter sysctl.conf to set some max segment sizes and such for your shared memory from past recall as well. Checking other clusters for examples.

**Comment by nariai** [comment]
For each job, you need to specify the memory that you are going to use by h_vmem. Let's say, STAR needs 8GB to load the genome index. If you want to run eight jobs in parallel, then you need to specify, for each job, h_vmem=8G (assuming if you are requesting one slot for each job). If you want to run eight jobs in parallel, and you want to use four threads for each job, then you need to specify -l h_vmem=2G -pe smp 4 for each job. By this way, each job requests 8GB memory in total. There is no concept of the "shared memory" in this cluster system. Correct me if my understanding is incorrect.

**Comment by tatarsky** [comment]
I assume he means shared memory ala `man shmget`. In my past dealings with STAR that is a way it operations. Your other comments I believe are correct however.

**Comment by tatarsky** [comment]
Bearing in mind h_vmem can't see or manage shared memory segments ala shmget....

**Comment by nariai** [comment]
If someone wants to run jobs in the short queue, please let me know. I can hold on my jobs.

**Comment by tatarsky** [comment]
I will mention the `qhold` command in passing (man qhold) as a nice way to self-hold your jobs. We have 90 cores in the all.q left and then the rules should be consistent.

**Comment by tatarsky** [comment]
And another item useful for "queue-sharing" is the array job `tc` flag which is discussed in `man submit`.

But a quick example of its use is as follows:

`qsub -t 1-100 -tc 10 (script)`

That says run my 100 entry array job 10 jobs at a time.

**Comment by cdeboever3** [comment]
@tatarsky, I think you said you have a script that can report the maximum memory usage of a process for the purpose of determining appropriate values of h_vmem. Can we get that, or can someone advise on the best way to figure this out? Thanks

**Comment by tatarsky** [comment]
I have a script I'd like to dust off/review before deploying. It came from Torque/Moab and I've had a couple of "it

Basically it uses the qacct -j command and parses out the mem and maxvmem output. While not perfect, it can give you some ideas of good targets.  Example:

```
mem          26.072GBs
maxvmem      7.703MB
```

**Comment by nariai** [comment]
Good point. I was always using the "top" commond to get the estimate of memory usage before submitting jobs. It would be good if we have a script to estimate that (memory requirement) easily.

**Comment by tatarsky** [comment]
To be clear, my script looking at the `qacct` which is a _post_ job run view of things. And assumes you've guessed over the level of memory needed and is showing you roughly a lower number compared to that. Otherwise it will show you your job got killed at the level of memory it tried to exceed.

It will not solve that initial guess. I often recommend a process similar to what you describe with top to get an initial guess or audit the malloc calls of the program if possible.

**Comment by tatarsky** [comment]
Its primary goal is _reducing_ your h_vmem setting following a "high bid"

**Comment by hurleyLi** [comment]
Do we get any alert message or error message if h_vmem break the job?
Also, it seems we don't get any "alert" if, say a short.q job run more than 2 days, coz I just found one of my jobs run on all.q quit without any "error message" after 48 hour run in stdout. 
Is there any way we could get such alert when jobs are kill by memory or running time? The only thing I can think of is using email alert.

**Comment by tatarsky** [comment]
All.q never had any h_vmem restrictions and they were only in affect for jobs submitted this morning. Nor was there any runtime limit. 

What is your job number? I suspect something else happened to it. 

As for alerts, the user has complete control over the alerts via the email method we discussed in #4 but I will elaborate on.

If you want alerts, consider the following either as a qsub or in the script

```
-m abe
```

Where "a" in the above is the item you seek to (a)lert on a failed job.
b=beginning
e=end.

**Comment by tatarsky** [comment]
I place the following into my submit scripts (with my email removed). I don't really care if a job tells me its begun (otherwise I could make that `abe`)

```
#$ -m ae
#$ -M (my email)
```

**Comment by hurleyLi** [comment]
Sorry I didn't keep the job ID for the failed job, and I already deleted to folder for re-run... 
The reason I suspect it's because of time limit, is that the job start at Nov  6 20:28:35, and the stderr file end at exactly Nov 8 20:28. 
I submitted this job to weekly queue now, and see whether I still get errors. 
And I added email alert to my jobs as well.

**Comment by tatarsky** [comment]
Hmm I would agree that looks like 48 hours. I wonder if it was in `short.q` somehow. I will be deleting all.q to prevent confusion in a moment and I'm trying to update the wiki with some details of todays efforts. 

I'll scan the logs though and see if I can determine what occurred. The weekly queue is set to have a runtime wallclock limit as follows:

```
s_rt                  168:00:00
h_rt                  168:10:00
```

**Comment by tatarsky** [comment]
Are you sure you are issuing jobs to the week.q?

I took a look at a few of your queued jobs and you appear to have `-l short` coming from somewhere. See the last line. Or are you doing both short and week items? 

```
job_number:                 1483
exec_file:                  job_scripts/1483
submission_time:            Mon Nov  9 17:18:50 2015
owner:                      hel070
uid:                        1042
group:                      users
gid:                        100
sge_o_home:                 /frazer01/home/hel070
sge_o_log_name:             hel070
sge_o_path:                 /opt/sge/bin:/opt/sge/bin/lx-amd64:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin:/frazer01/home/hel070/bin:/frazer01/projects/CARDIPS/pipeline/WGS/script
sge_o_shell:                /bin/bash
sge_o_workdir:              /frazer01/projects/CARDIPS/pipeline/WGS/alignment_bwa
sge_o_host:                 fl-hn1
account:                    sge
stderr_path_list:           NONE:NONE:/frazer01/projects/CARDIPS/pipeline/WGS/alignment_bwa/94f9d79c-a7f6-4b0a-bc0a-869468185844/wgs_mdup.err
hard resource_list:         h_vmem=4G,short=TRUE
```

**Comment by hurleyLi** [comment]
Actually the last batch of submission, which happened 15 minutes ago, is the first time I used -l short. I believe I used default queue for all previous jobs.

**Comment by tatarsky** [comment]
OK. Just checking. You know that will limit those jobs to 48 hours is what I'm confirming.

**Comment by hurleyLi** [comment]
yes. they should really finished within 48 hours. That failed one is a odd case, I guess... Since no one is using the short queue now, so I'm taking it :)

**Comment by tatarsky** [comment]
Gotcha. I'm poking around to try to explain that failure. But also wanting to capture some of the items today into the wiki ;)

**Comment by tatarsky** [comment]
So I did find that job (548) and yes it did get killed. I'm not sure why but I'm guessing it may have gotten caught in my efforts Friday to make the short.q. I'm looking at your remaining all.q jobs from the point of view of their shells and they don't seem to have limits. But let me know if you see this again. 

I show two jobs left in all.q and then thats it for that queue:

```
   956 0.55617 WGS_alignA hel070       r     11/08/2015 07:27:49 all.q@fl-n-1-2                    10        
    958 0.55617 WGS_alignA hel070       r     11/08/2015 08:15:19 all.q@fl-n-1-2                    10       
```

Here is job 956 from the point of view of its running shell on Node 2 where its pid 23887

```
cat /proc/23887/limits
Max cpu time              unlimited            unlimited            seconds   
```

And here is job 958:

```
Max cpu time              unlimited            unlimited            seconds   
```

So I believe they will complete with no limit.

**Comment by tatarsky** [comment]
How we doing out there today in SGE land? I show good solid load. Everybody ok so far with the rules and queues? NO changes today...just watching.

**Comment by nariai** [comment]
Looks good, working as expected.

**Comment by hurleyLi** [comment]
Everything looks normal! If I have to ask anything, it seems that week.q is using 390 cores in total, rather than the limit 384 cores. Maybe because some of the jobs start to run before the week queue is implemented yesterday (from 08:13)? I'm not sure...  But the jobs after that are waiting on the queue, which is expected.

**Comment by tatarsky** [comment]
I'll double check but I did add that quota I believe with some jobs already running. I'll double check it settles to the value we expect.

**Comment by cdeboever3** [comment]
@tatarsky, we met with @kafrazer today and decided on the following changes to the queue set up based on historical usage of the old cluster and anticipated usage of the new one:
1. Create new queue all.q which can use any node with any number of slots (maximum 512 slots)
2. Maximum wall time of the all.q will be two days
3. Current default queue will be renamed week
4. The all.q will have the lower priority than others (short, week, and long) when waiting in the queue
5. The all.q will be the default queue

We also discussed migrating the old cluster nodes into the new cluster. We figure that migrating eight of the opteron nodes first as a test and to ease the transition for everyone else makes sense. When migrating the eight Opteron nodes (as the first phase) into the new cluster:
1. The all2.q will be set up, which will only use the eight Opteron nodes (maximum 256 slots)
2. Maximum wall time of the all2.q will be seven days
3. Other queues (all, short, week, and long) will not use the Opteron nodes

We can clear up any queue questions on the call tomorrow if need be. We all agree that we don't want to change the queue settings very often so we are just trying to nail down something that should work for a while. Thanks

**Comment by nariai** [comment]
I guess, since the default queue is now the week.q, please ignore the point 3 above. The new "all.q" will be the default queue. We will speficy "-l week" in order to submit jobs in the week queue (which has the higher priority than "all.q") after the change.

**Comment by tatarsky** [comment]
My only comment is "short" and "all" are roughly the same thing in terms of maximum walltime. Do you care? Otherwise the above is all doable and sounds fine. The key question is "when would you like this done"

**Comment by hurleyLi** [comment]
Hi @tatarsky , just to clarify, if I turn on -pe smp, then -l h_vmem specifies memory for each core, not total core, right?

**Comment by tatarsky** [comment]
I believe that is correct.

**Comment by cdeboever3** [comment]
Hi all, Paul is going to reactivate the all queue tomorrow with lower
priority than short, week, and all as discussed. all will become the new
default and have 2 days walltime limit so please be ready for that change.

**Comment by hurleyLi** [comment]
Will the jobs running currently be affected?

Hurley

Sent from my iPhone

On Nov 11, 2015, at 2:18 PM, Christopher DeBoever <notifications@github.com<mailto:notifications@github.com>> wrote:

Hi all, Paul is going to reactivate the all queue tomorrow with lower
priority than short, week, and all as discussed. all will become the new
default and have 2 days walltime limit so please be ready for that change.

**Comment by tatarsky** [comment]
Not to my knowledge. The main change will be that the `week.q` will require selection with `qsub -l week` or by default items will go into the `all.q`.

**Comment by tatarsky** [comment]
`all.q` was re-enabled as the default queue with a 48 hour maximum walltime.
`week.q` now requires `qsub -l week (script)` to be placed in it.

Working on making sure the priority items are as described. No other changes yet this morning.

**Comment by tatarsky** [comment]
I took the limit off of `all.q` because several jobs were queued without `-l week` in them and I don't think that is the intent. Please @hurleyLi you might check that your jobs running now in `all.q` how far they are and you may wish to restart them with `-l week`

**Comment by tatarsky** [comment]
I've disabled `all.q` until the queued jobs are truly ready for it because I suspect the @hurleyLi jobs want `-l week`. Please let me know when that is ready as I don't want items in `all.q` that were not intended there. At the moment this means there is NO ENABLED DEFAULT QUEUE. So please advise when I can enable it and not have items drop into it unplanned.

**Comment by hurleyLi** [comment]
Hi @tatarsky, I deleted the jobs waiting on the queue, you may implement the change anytime, and I'll re-submit those jobs later. Thanks!
Also I'm going to submit jobs on the all.q. The intention of this change is that people don't have to manually split the jobs to both week.q and short.q, if they want to use all of the nodes. And if other people want to submit jobs while all.q is full, they'll submit to the short.q, since short.q will have higher priority.

**Comment by tatarsky** [comment]
Above sounds correct. I may not have all the priorities right yet but I'll get `all.q` enabled. Just remember its got a 48 hour wallclock limit and thats all I was concerned about for you....

**Comment by tatarsky** [comment]
`all.q` re-enabled. I'm double checking priorities and such in a moment but the default queue is now back with a 48 hour runtime.

week.q and short.q are also as before only week.q require `-l week` to enter.

**Comment by tatarsky** [comment]
I believe people are submitting OK and I'm showing what I believe is correct priorities. I am double checking a few items but not planning any changes today. But please let me know if you feel anything is amiss.

**Comment by hurleyLi** [comment]
Hi @tatarsky, it looks like the priority is not set as what we expected. Or maybe there's some weird things happened for array jobs. When we submit single job on the short.q, it's not always coming in front of the all.q jobs on queue. Maybe because the the jobs on all.q are array jobs? And another thing I noticed is, it seems the more cores we request, the higher priority it get. Could you please clarify that? 

Thanks,

**Comment by tatarsky** [comment]
I'll look in the morning.

I believe for array jobs SGE assigns some additional priority trying to keep an array job that has components already running. I believe its tunable. (Will check, I'm not actually looking at it)

And the core item is also I believe tunable with SGE there trying to help multiple core jobs from being "queue starved" by lots of small jobs. 

But basically I'll look into it.

**Comment by tatarsky** [comment]
Also note I'll activate a `qstat -pri` loop so we can look at the various options as we tune. Just note a time when you observe something to make it a bit easier.

**Comment by tatarsky** [comment]
So we have a MANY items we can consider in various ways but lets do things one at a time so we can make sure its what you want. Also, in the above comment `qconf -pri -urg` which shows more of the values used to derive a queued jobs priority is tracking the queue state while we tune so you only need to provide a time for me to look at the scores. 

IMPORTANT: I corrected this morning the `qsub` manpage which did not properly get installed. There is a stub POSIX qsub manpage in many distros and its not correct and will now no longer appear. 

`man qsub` should now return the proper SGE specific page which should start with:

```
SUBMIT(1)                  Grid Engine User Commands                 SUBMIT(1)

NAME
       qsub, qsh, qlogin, qrsh, qalter, qresub - submit jobs to Grid Engine
```

/// User control priorities

First off, I can remind and document the ability by users to up their jobs priority regardless of queue with the `-p X` flag to a qsub. 

The default qsub priority has been set to `-100` to allow folks to relatively rank and also to `qalter` items on their own. Its slightly annoying in that priorities for regular users go from 0 -> -1024 so the number "0" is considered the highest for a user (I can define SGE managers capable of assigning priorities greater than 0 at some point when ready)

`qsub -p 0 foo.sh`

That will up the priority regardless of queue. `qalter` can also be used to adjust that while queued without resubmit. For example here making it less than the default priority.

`qalter -p -200 (jobid)`

/// Short Resource Urgency score tuning

I can add some additional urgency score for the short queue which I am going to do NOW to make sure its score is getting high enough. Requesting the `short` resource now gets an immediate 10000 points per slot. Which should trump just about anything currently configured. Please test further. 

/// Urgency per slot

Then, I can REDUCE the urgency score assign based on the number of processors but remember I am pretty sure that is part of SGEs desire to not have such jobs get sidelined in the queue by lots of small low slot count jobs. We can tune it to workload a bit. For my memory that value right now is 1000 per slot. I propose we reduce it to 500 per slot if the above doesn't sort the way you want. 

/// Array jobs and their calculations

Array jobs are I believe indeed treated as one priority calculation by design. You can self-limit an arrays running components with the `-tc XX` number. Such as:

`qsub -t 1-100 -tc 20 foo.sh`

Which means here is an array job with 100 items but please SGE only run 20 at a time. 

/// Other things we do not have tuned on at all

SGE has a couple of other simple items for queue priorities before you start getting into fairshare trees. Those include: deadline setting and queue wait time re-prioritizing. Deadline setting can be done with the `qsub -dl YY]MMDDhhmm` format. But is not currently turned on. 

Let me know how things go with the one change and other information here. Happy Friday.

**Comment by hurleyLi** [comment]
Thanks @tatarsky! These new features will be very convenient, and we'll give it a try.

**Comment by tatarsky** [comment]
Bottom line SGE has an absurd number of ways to do things. So if you keep describing when you don't think its what you want mark the time down if possible to catch an example and update this! 

While I don't claim to know all the ways to solve various items, SGE remains source code and can often be consulted for reasons something does something.

**Comment by tatarsky** [comment]
I am showing a good example of the upped priorities right now with short.q jobs passing some array items. Would you agree that is looking good?

```
 2772 0.55617 1.00000 0.45117 0.50000  -100 pp         hiroko       r     11/23/2015 14:17:56 short.q@fl-n-1-15                  4 17
   2772 0.55617 1.00000 0.45117 0.50000  -100 pp         hiroko       r     11/23/2015 14:20:26 short.q@fl-n-1-16                  4 18
   2772 0.55617 1.00000 0.45117 0.50000  -100 pp         hiroko       r     11/23/2015 14:22:26 short.q@fl-n-1-16                  4 19
   2772 0.55617 1.00000 0.45117 0.50000  -100 pp         hiroko       r     11/23/2015 14:22:26 short.q@fl-n-1-16                  4 20
   2772 0.55617 1.00000 0.45117 0.50000  -100 pp         hiroko       r     11/23/2015 14:22:41 short.q@fl-n-1-15                  4 21
   2772 0.55617 1.00000 0.45117 0.50000  -100 pp         hiroko       r     11/23/2015 14:37:11 short.q@fl-n-1-15                  4 22
   2772 0.55617 1.00000 0.45117 0.50000  -100 pp         hiroko       qw    11/23/2015 12:02:08                                    4 23-50:1
   2773 0.55617 1.00000 0.45117 0.50000  -100 pp         hiroko       qw    11/23/2015 12:02:50                                    4 1-20:1
   2774 0.51500 0.10000 0.50000 0.50000     0 bowtie2_01 hiroko       qw    11/23/2015 14:37:07                                    8 1
   2676 0.45617 0.00000 0.45117 0.50000  -100 WGS_GATK_b hel070       qw    11/23/2015 09:37:26                                    4 3-22:1
   2677 0.45617 0.00000 0.45117 0.50000  -100 WGS_GATK_b hel070       qw    11/23/2015 09:37:26                                    4 1-22:1
   2678 0.45617 0.00000 0.45117 0.50000  -100 WGS_GATK_b hel070       qw    11/23/2015 09:37:26                                    4 1-22:1
```

**Comment by hirokomatsui** [comment]
Yes, it's working pretty well!

**Comment by tatarsky** [comment]
OK. Good. Noting the changes! I will not further downgrade the urgent points for multiple cores at this time.

**Comment by tatarsky** [comment]
OK. I am going to place there here for comment. I believe that we want all the old nodes in a completely separate non-default queue. Do we have a name preference for that queue? I may get to this shortly depending on load.

**Comment by nariai** [comment]
Yes, we decided to name the non-default queue as "all2".

**Comment by nariai** [comment]
Or, if people think there is a better name for it, I'm fine with changing the name.

**Comment by cdeboever3** [comment]
Maybe name them after their processors (are they opterons or something?)
and perhaps have a 1g on there to remind people of the networking
difference? Something like opt1g? As you've mentioned, names like "old"
tend to decrease popularity.

**Comment by nariai** [comment]
"opteron" sounds better than "opt1g", just in case we can connect them within 10g network in the future.

**Comment by tatarsky** [comment]
The Lustre as provided appears to be only on the 10G. I am asking Advanced HPC to ask Intel for the proper steps to add it to the 1G. So at the moment we are waiting on that. More information as I get it.

**Comment by tatarsky** [comment]
While I wait for what I suspect will be a fairly complex process I am going to proceed with the NFS mount of the `frazer01` Lustre via the second head node. Given all this I/O is being forced via a single 1G link anyway (the connection between the racks) I feel this won't be a initial huge bottle neck.

I feel at least you will be able to get some compute jobs running. 

I think we should discuss the refurb of the old cluster with 10G as a more performance ideal and less complex path. 

I will still pursue the steps to add lnet support to the 1G side.

**Comment by tatarsky** [comment]
OK. So for this evening I've gotten some jobs running on `cn1` using a queue called for now `opt.q`. This most mostly as a proof of concept the NFS would get us moving along and it does.

**Comment by tatarsky** [comment]
More details tomorrow.

**Comment by cdeboever3** [comment]
Cool. I have some compute heavy, IO light jobs that I can send to them when
the queue is ready.

**Comment by tatarsky** [comment]
If you wish to perform a simple test with a single node so far use the `opt` resource to request the `opt.q`.

```
qsub -l opt (some item)
```

So far that will just be `cn1`

I've not made memory a consumable or any other walltime limits yet on this queue and I clearly state not all the RPM loads are finished. I am working on the other nodes but focusing on automation.

**Comment by tatarsky** [comment]
Old nodes except for `cn7` and `cn12` are in SGE `opt.q` which is selected with the `-l opt` resource. Report issues in #33 and I'll get to them fairly quickly but with some in/out today for me.

Remember, its a single 1G link over to these nodes and for now we are using an NFS->Lustre method.

**Comment by tatarsky** [comment]
So this is to remind me: I have NOT turned on memory as a consumable yet on `opt.q`. I will shortly after an errand. I am going to start with an oversubsciption of 74G (real=54G"

There is no walltime limit on the `opt.q` either. I assume for now that is fine.

**Comment by tatarsky** [comment]
Walltime on `opt.q` set to one week per above and memory as a consumable activated with some oversubscription.

**Comment by tatarsky** [comment]
The SGE s_core limit (soft core limit) was changed from UNLIMITED to 0 a moment ago per #102 

This means the default UNIX limit "coredump" is set to zero which results in a crashing program not dumping potentially a large core file. A recent reminder of the joys of this with many large memory jobs all crashing due to a bug was the impetus for this what I normally do as a default and forgot. 

If you NEED coredumps in a job add a `ulimt -c unlimited` to your submit script.

**Comment by nariai** [comment]
I submitted a group of jobs (job id: 405730-405751) by
qsub -l long
However, it looks like only two of the submitted jobs are using long.q, whereas other jobs are using all.q.

I'm expecting that the submitted jobs can be running more than two days (that is why I submitted jobs in long queue), but it looks like some of the jobs running in all.q will be terminated after two days? Is this situation (submit jobs in long.q, but actually using all.q) expected? Do you have any ideas?

**Comment by tatarsky** [comment]
I'm looking at job 405731 as an example of this.

No, I would say that is not expected but will look at the logs to see what I can locate to explain it or locate a config item that needs to change.

**Comment by tatarsky** [comment]
And if you have handy the precise qsub command you issued it would be helpful.

**Comment by tatarsky** [comment]
Assuming it wasn't just `qsub -l long calc_LD_005_ALL.sh` BTW...just checking if any other command line args were used...

**Comment by nariai** [comment]
Thank you for taking a look.

The command was (in Makefile):

calc_LD_005_ALL:
    number=1 ; while [[ $$number -le 22 ]] ; do \
        qsub -l long -p 0 -pe smp 1 -cwd -e logs -o stdouts -v ARG1=$$number calc_LD_005_ALL.sh ; \
        ((number = number + 1)) ; \
    done

The path is:
/projects/T2D/analysis/common_variants

**Comment by tatarsky** [comment]
OK. I noted that "pe smp 1" in and wanted to check on that. Is there a reason for requesting a PE of only one processor? (compared to more than one as that is really the same as not asking for a PE). Shouldn't matter but I've actually never used that construct. 

I will try to reproduce using a similar makefile. I wonder if the `-l long` is somehow not being fully preserved in the make commands that are finally issued.

I wonder if the behavior will change if you embed the SGE hard resource in the script ala:

```
#$ -l long
```

**Comment by tatarsky** [comment]
I have reproduced the above. For whatever reason the first job is issued to long.q as expected but the following jobs are in all.q

I am trying to step through the outputted items to see why.

**Comment by nariai** [comment]
When I changed from "-l long" to "-l week", all of the jobs are running in week.q (405752-405773).
Other parameters were set exactly the same.
Is this something long.q specific?

**Comment by tatarsky** [comment]
Interesting. When I removed "pe smp 1" they all ended up in long.q ;)

So I am suspecting something about the final commands being output here.

Is there a reason you are doing this make loop compared to an SGE array job?

**Comment by nariai** [comment]
Very interesting. No, I didn't have a specific reason not using an array job. Next time, I'll try to remove "pe smp 1" if I run jobs with one cpu, or try to use an array job.

**Comment by tatarsky** [comment]
Let me look at it a bit more in the morning. Whatever is going on its subtle. 

The array job suggestion I can explain more its just a simpler way of running a series of jobs from the same submit script using SGE built in shell variables that increment similar to your make file loop. Basically a notch less setup work and SGE is a notch more efficient with array jobs particularly for large arrays. 

But I'd still like to understand what is going on there. 

I am suspicious of the PE perhaps confusing things but I don't know why it would drop into all.q. 

So "unsure" is the answer for now and I will look closer with some coffee in the morning.

**Comment by nariai** [comment]
Thanks, please let me know if you can figure it out in more detail. I just re-submitted jobs in long queue.

**Comment by tatarsky** [comment]
So basically its got something to do with the `-pe smp 1` but I don't know why. Tracing the make file shows clearly that your qsubs are as expected but that under some circumstance its like the `-l long` is dropped. 

Removing the `-pe smp 1` and it works fine. Trying to place the `-l long` into the submit script does NOT change the outcome. Changing the ordering does NOT change the outcome. 

```
qsub -l long -pe smp 1 -p 0 -cwd -e logs -o stdouts -v ARG1=17 foo.sh
```

So I'm suspecting some kind of qsub argument bug at the moment but will double check my queue assumptions. 

I am also going to try different `pe smp` options. Technically "pe smp 1" is not really a concept you need to specify but I would assume even if you did SGE wouldn't care.

**Comment by tatarsky** [comment]
Interestingly using `-pe smp 2` results in everything going into `long.q` fine. At least for my tests. 

So I'm guessing a form of bug involving asking for an SMP environment (multiple processors) but then really saying "not really" as that is what "1" would mean.

**Comment by tatarsky** [comment]
There are a few mutterings on the mailing lists of somewhat similar experiences. But nothing conclusive. I believe the queues are correct but that the outcome when the PE is `1` is probably a bug. I'll look a little more but my belief is also that stanza isn't ever needed (asking for one processor slot).

**Comment by tatarsky** [comment]
BTW, random choices of PE smp values _above_ 1 all seem to result in the jobs correctly in the long.q.
So I'm guessing more its a bug.

**Comment by tatarsky** [comment]
SGE spool moved to `fl-ims` per #178 to reduce impact if `fl-hn1` crashes during attempts to determine reason. `fl-hn2` shadow qmaster now will not hang on NFS spool. No cross head node dependencies.

**Comment by tatarsky** [comment]
Moved seq_no for juplow and juphigh to non-matching seq_no's as other queues.
Based on possible assignment to said queues under some cases of jobs not referencing their required resource. #188 

Current settings for reference:
```
short     0,[@notlonghosts=0],[@longhosts=5]
juplow    1,[@notlonghosts=1],[@longhosts=6]
juphigh   2,[@notlonghosts=2],[@longhosts=7]
opt       3
week      5,[@notlonghosts=5],[@longhosts=10]
all       10,[@notlonghosts=10],[@longhosts=15
long      10
```

**Comment by tatarsky** [comment]
Just for reference. `c7.q` was added. Single host in hostgroup for it `cn12`. queue resource `c7`. Defaults for everything else.

**Comment by tatarsky** [comment]
Make a "jupsmp" PE environment in prep for possible fix for #188 
Remove `make` and `mpi` from juplow.q and juphigh.q as never needed for those queues and in theory related to proposed solution to #188 (although we don't see much mpi or make PE on this cluster)

**Comment by tatarsky** [comment]
For reference: Per Git #278 the s_rt and h_rt were removed from juphigh.q.

**Comment by tatarsky** [comment]
`opt_q` and all `cn*` nodes will be removed shortly. Made a backup of SGE config and will start removing. Already removed them as an option in JupyterHubs.

**Comment by tatarsky** [comment]
Cluster shutdown.
